In [ ]:
import numpy as np 
import pandas as pd 

In [3]:
data= pd.read_csv('MUP_DPR_RY25_P04_V10_DY23_Geo.csv', sep = ',')

data = pd.DataFrame(data)

data

C:\Users\HP\AppData\Local\Temp\ipykernel_7888\243678716.py:1: DtypeWarning: Columns (1) have mixed types. Specify dtype option on import or set low_memory=False.
  data= pd.read_csv('MUP_DPR_RY25_P04_V10_DY23_Geo.csv', sep = ',')


Prscrbr_Geo_Lvl Prscrbr_Geo_Cd Prscrbr_Geo_Desc  \
0             National            NaN         National   
1             National            NaN         National   
2             National            NaN         National   
3             National            NaN         National   
4             National            NaN         National   
...                ...            ...              ...   
115931           State             9E  Foreign Country   
115932           State             9E  Foreign Country   
115933           State             9E  Foreign Country   
115934           State             9E  Foreign Country   
115935           State             9E  Foreign Country   

                            Brnd_Name                       Gnrc_Name  \
0            1st Tier Unifine Pentips            Pen Needle, Diabetic   
1       1st Tier Unifine Pentips Plus            Pen Needle, Diabetic   
2                            Abacavir                Abacavir Sulfate   
3                 Abacavir-Lamivudine     Abacavir Sulfate/Lamivudine   
4                             Abilify                    Aripiprazole   
...                               ...                             ...   
115931                      Xigduo Xr  Dapaglifloz Propaned/Metformin   
115932                         Xtandi                    Enzalutamide   
115933                        Yuvafem                       Estradiol   
115934                Ziprasidone Hcl                 Ziprasidone Hcl   
115935              Zolpidem Tartrate               Zolpidem Tartrate   

        Tot_Prscrbrs  Tot_Clms  Tot_30day_Fills  Tot_Drug_Cst  Tot_Benes  ...  \
0                691      1613           2874.2      44355.04      699.0  ...   
1               1167      3269           5592.0      97951.18     1267.0  ...   
2               2617     19634          25152.1    5290175.88     2807.0  ...   
3               2329     20191          27901.0    9981620.25     2890.0  ...   
4               2111     12096          15928.9   11867897.88     1734.0  ...   
...              ...       ...              ...           ...        ...  ...   
115931             2        17             25.0      11384.92        NaN  ...   
115932             1        29             29.0     149316.49        NaN  ...   
115933             3        72            142.0       6466.87       20.0  ...   
115934             1        31             33.0       3994.67        NaN  ...   
115935            11       116            138.0       1186.16       43.0  ...   

       GE65_Tot_30day_Fills  GE65_Tot_Drug_Cst  GE65_Bene_Sprsn_Flag  \
0                    2375.6           35803.14                   NaN   
1                    4545.0           76544.14                   NaN   
2                   15353.6         3042569.00                   NaN   
3                   18282.9         6206789.76                   NaN   
4                    4897.1         3318969.72                   NaN   
...                     ...                ...                   ...   
115931                  NaN                NaN                     *   
115932                 29.0          149316.49                     *   
115933                142.0            6466.87                   NaN   
115934                  0.0               0.00                     #   
115935                105.0             785.58                     #   

        GE65_Tot_Benes LIS_Bene_Cst_Shr  NonLIS_Bene_Cst_Shr  \
0                565.0           992.77              7093.55   
1               1035.0          1899.59             10392.03   
2               1662.0         10643.29            152140.95   
3               1877.0         14925.90            360305.92   
4                578.0         25007.42            178277.87   
...                ...              ...                  ...   
115931             NaN            17.20               750.44   
115932             NaN             0.00                 0.00   
115933            2

- Remove duplicate

In [4]:
# Find duplicated rows (entire row is the same)
duplicates = data[data.duplicated()]
print(duplicates)


Empty DataFrame
Columns: [Prscrbr_Geo_Lvl, Prscrbr_Geo_Cd, Prscrbr_Geo_Desc, Brnd_Name, Gnrc_Name, Tot_Prscrbrs, Tot_Clms, Tot_30day_Fills, Tot_Drug_Cst, Tot_Benes, GE65_Sprsn_Flag, GE65_Tot_Clms, GE65_Tot_30day_Fills, GE65_Tot_Drug_Cst, GE65_Bene_Sprsn_Flag, GE65_Tot_Benes, LIS_Bene_Cst_Shr, NonLIS_Bene_Cst_Shr, Opioid_Drug_Flag, Opioid_LA_Drug_Flag, Antbtc_Drug_Flag, Antpsyct_Drug_Flag]
Index: []

[0 rows x 22 columns]


In [5]:
# check dataset info 

data.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 115936 entries, 0 to 115935
Data columns (total 22 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   Prscrbr_Geo_Lvl       115936 non-null  object 
 1   Prscrbr_Geo_Cd        112328 non-null  object 
 2   Prscrbr_Geo_Desc      115935 non-null  object 
 3   Brnd_Name             115936 non-null  object 
 4   Gnrc_Name             115936 non-null  object 
 5   Tot_Prscrbrs          115936 non-null  int64  
 6   Tot_Clms              115936 non-null  int64  
 7   Tot_30day_Fills       115936 non-null  float64
 8   Tot_Drug_Cst          115936 non-null  float64
 9   Tot_Benes             94340 non-null   float64
 10  GE65_Sprsn_Flag       21063 non-null   object 
 11  GE65_Tot_Clms         94873 non-null   float64
 12  GE65_Tot_30day_Fills  94873 non-null   float64
 13  GE65_Tot_Drug_Cst     94873 non-null   float64
 14  GE65_Bene_Sprsn_Flag  48848 non-null   object 
 15  

In [6]:
# check dataset cols 
data.columns

Index(['Prscrbr_Geo_Lvl', 'Prscrbr_Geo_Cd', 'Prscrbr_Geo_Desc', 'Brnd_Name',
       'Gnrc_Name', 'Tot_Prscrbrs', 'Tot_Clms', 'Tot_30day_Fills',
       'Tot_Drug_Cst', 'Tot_Benes', 'GE65_Sprsn_Flag', 'GE65_Tot_Clms',
       'GE65_Tot_30day_Fills', 'GE65_Tot_Drug_Cst', 'GE65_Bene_Sprsn_Flag',
       'GE65_Tot_Benes', 'LIS_Bene_Cst_Shr', 'NonLIS_Bene_Cst_Shr',
       'Opioid_Drug_Flag', 'Opioid_LA_Drug_Flag', 'Antbtc_Drug_Flag',
       'Antpsyct_Drug_Flag'],
      dtype='object')

In [7]:
# show a description for the dataset 
data.describe()

Tot_Prscrbrs      Tot_Clms  Tot_30day_Fills  Tot_Drug_Cst  \
count  115936.000000  1.159360e+05     1.159360e+05  1.159360e+05   
mean     1633.927667  2.789245e+04     5.087556e+04  4.753530e+06   
std     14917.992926  4.702736e+05     1.019406e+06  8.787711e+07   
min         1.000000  1.100000e+01     1.100000e+01  1.506000e+01   
25%        14.000000  5.100000e+01     6.870000e+01  1.791928e+04   
50%        62.000000  2.830000e+02     3.863000e+02  1.291858e+05   
75%       427.000000  2.503000e+03     3.623925e+03  8.348181e+05   
max    725303.000000  6.846728e+07     1.679827e+08  1.826613e+10   

          Tot_Benes  GE65_Tot_Clms  GE65_Tot_30day_Fills  GE65_Tot_Drug_Cst  \
count  9.434000e+04   9.487300e+04          9.487300e+04       9.487300e+04   
mean   1.030317e+04   2.822537e+04          5.346846e+04       4.423261e+06   
std    1.412452e+05   4.508361e+05          1.011888e+06       8.345800e+07   
min    1.100000e+01   0.000000e+00          0.000000e+00       0.000000e+00   
25%    3.800000e+01   6.100000e+01          8.050000e+01       1.969902e+04   
50%    1.690000e+02   3.820000e+02          5.229000e+02       1.345428e+05   
75%    1.284000e+03   3.074000e+03          4.619300e+03       7.773636e+05   
max    1.677038e+07   6.090046e+07          1.521763e+08       1.716443e+10   

       GE65_Tot_Benes  LIS_Bene_Cst_Shr  NonLIS_Bene_Cst_Shr  
count    6.708800e+04      1.159360e+05         1.159360e+05  
mean     1.238629e+04      1.167093e+04         3.127506e+05  
std      1.475074e+05      1.765071e+05         6.623784e+06  
min      0.000000e+00      0.000000e+00         0.000000e+00  
25%      7.800000e+01      2.155000e+01         1.113705e+03  
50%      3.590000e+02      1.860150e+02         8.818410e+03  
75%      2.212250e+03      1.938015e+03         6.111388e+04  
max      1.509227e+07      3.274342e+07         1.712191e+09

- Standarization 

In [8]:
# we will disply all the object data 
# after we displyed the obejct cols , we will disply the data of each col

object_data = data.select_dtypes(include=['object']).columns
print(object_data)

Index(['Prscrbr_Geo_Lvl', 'Prscrbr_Geo_Cd', 'Prscrbr_Geo_Desc', 'Brnd_Name',
       'Gnrc_Name', 'GE65_Sprsn_Flag', 'GE65_Bene_Sprsn_Flag',
       'Opioid_Drug_Flag', 'Opioid_LA_Drug_Flag', 'Antbtc_Drug_Flag',
       'Antpsyct_Drug_Flag'],
      dtype='object')


- This code it you want to check all the data or just on signle col

In [9]:
# Show all rows
pd.set_option('display.max_rows', None)

# Show all columns
pd.set_option('display.max_columns', None)

# Optional: widen display for better viewing in notebooks
pd.set_option('display.width', 1000)

# Now show the full DataFrame or a portion
print(data['Prscrbr_Geo_Lvl'])  # Show first 100 rows

# its very a loot of lines to check we can just use 

0         National
1         National
2         National
3         National
4         National
5         National
6         National
7         National
8         National
9         National
10        National
11        National
12        National
13        National
14        National
15        National
16        National
17        National
18        National
19        National
20        National
21        National
22        National
23        National
24        National
25        National
26        National
27        National
28        National
29        National
30        National
31        National
32        National
33        National
34        National
35        National
36        National
37        National
38        National
39        National
40        National
41        National
42        National
43        National
44        National
45        National
46        National
47        National
48        National
49        National
50        National
51        National
52        Na

In [10]:
# check if the col is empty or no 
data['Prscrbr_Geo_Lvl'].notna().any()  # → True if column has data

True

In [11]:
# we will check the col catg 
print(data['Prscrbr_Geo_Lvl'].unique())
# now we now that the col Prscrbr_Geo_Lvl conatine 2 categ {National , State} 
# why we did this to check if the col containe any error

['National' 'State']


In [12]:
# check if the col is empty or no 
data['Prscrbr_Geo_Cd'].notna().any()  # → True if column has data

True

In [13]:
# we will check the col catg 
print(data['Prscrbr_Geo_Cd'].unique())
# why we did this to check if the col containe any error

[nan 1.0 2.0 4.0 5.0 6.0 8.0 9.0 10.0 11.0 12.0 13.0 15.0 16.0 17.0 18 19
 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 44
 45 46 47 48 49 '49' '50' '51' '53' '54' '55' '56' '60' '66' '69' '72'
 '78' '9A' '9B' '9C' '9D' '9E']


In [14]:
 
# check if the col is empty or no 
data['Prscrbr_Geo_Desc'].notna().any()  # → True if column has data

True

In [15]:
# we will check the col catg 
print(data['Prscrbr_Geo_Desc'].unique())
# why we did this to check if the col containe any error

['National' nan 'Alabama' 'Alaska' 'Arizona' 'Arkansas' 'California'
 'Colorado' 'Connecticut' 'Delaware' 'District of Columbia' 'Florida'
 'Georgia' 'Hawaii' 'Idaho' 'Illinois' 'Indiana' 'Iowa' 'Kansas'
 'Kentucky' 'Louisiana' 'Maine' 'Maryland' 'Massachusetts' 'Michigan'
 'Minnesota' 'Mississippi' 'Missouri' 'Montana' 'Nebraska' 'Nevada'
 'New Hampshire' 'New Jersey' 'New Mexico' 'New York' 'North Carolina'
 'North Dakota' 'Ohio' 'Oklahoma' 'Oregon' 'Pennsylvania' 'Rhode Island'
 'South Carolina' 'South Dakota' 'Tennessee' 'Texas' 'Utah' 'Vermont'
 'Virginia' 'Washington' 'West Virginia' 'Wisconsin' 'Wyoming'
 'American Samoa' 'Guam' 'Northern Mariana Islands' 'Puerto Rico'
 'Virgin Islands' 'Armed Forces Central/South America'
 'Armed Forces Europe' 'Armed Forces Pacific' 'Unknown' 'Foreign Country']


In [16]:
# check if the col is empty or no 
data['Brnd_Name'].notna().any()  # → True if column has data

True

In [17]:
# we will check the col catg 
print(data['Brnd_Name'].unique())
# why we did this to check if the col containe any error

['1st Tier Unifine Pentips' '1st Tier Unifine Pentips Plus' 'Abacavir' ...
 'Zyprexa Zydis' 'Zytiga' 'Zyvox']


In [18]:
# check if the col is empty or no 
data['Gnrc_Name'].notna().any()  # → True if column has data

True

In [19]:
# we will check the col catg 
print(data['Gnrc_Name'].unique())
# why we did this to check if the col containe any error

['Pen Needle, Diabetic' 'Abacavir Sulfate' 'Abacavir Sulfate/Lamivudine'
 ... 'Loncastuximab Tesirine-Lpyl' 'Pitavastatin Magnesium'
 'Olanzapine Pamoate']


In [20]:
# we will check the col catg 
print(data['GE65_Sprsn_Flag'].unique())
# why we did this to check if the col containe any 


[nan '*' '#']


In [21]:
# Delate the col GE65_Sprsn_Flag it has just {nan,*,#}

data.drop('GE65_Sprsn_Flag', axis=1, inplace=True)

In [22]:

# check if the col is empty or no 
data['GE65_Bene_Sprsn_Flag'].notna().any()  # → True if column has data

True

In [23]:
# we will check the col catg 
print(data['GE65_Bene_Sprsn_Flag'].unique())
# why we did this to check if the col containe any 

[nan '*' '#']


In [24]:
# delete the col GE65_Bene_Sprsn_Flag 
data.drop('GE65_Bene_Sprsn_Flag',axis=1, inplace=True)

In [25]:

# check if the col is empty or no 
data['Opioid_Drug_Flag'].notna().any()  # → True if column has data

True

In [26]:
# we will check the col catg 
print(data['Opioid_Drug_Flag'].unique())
# why we did this to check if the col containe any 

['N' 'Y']


In [27]:
# check if the col is empty or no 
data['Opioid_LA_Drug_Flag'].notna().any()  # → True if column has data

True

In [28]:
# we will check the col catg 
print(data['Opioid_LA_Drug_Flag'].unique())
# why we did this to check if the col containe any

['N' 'Y']


In [29]:

# check if the col is empty or no 
data['Antbtc_Drug_Flag'].notna().any()  # → True if column has data

True

In [30]:
# we will check the col catg 
print(data['Antbtc_Drug_Flag'].unique())
# why we did this to check if the col containe any

['N' 'Y']


In [31]:

# check if the col is empty or no 
data['Antpsyct_Drug_Flag'].notna().any()  # → True if column has data

True

In [32]:
# we will check the col catg 
print(data['Antpsyct_Drug_Flag'].unique())
# why we did this to check if the col containe any

['N' 'Y']


In [33]:
# Also we going to check the numerical cols 
numeric = data.select_dtypes(include='number').columns
print(numeric)


Index(['Tot_Prscrbrs', 'Tot_Clms', 'Tot_30day_Fills', 'Tot_Drug_Cst', 'Tot_Benes', 'GE65_Tot_Clms', 'GE65_Tot_30day_Fills', 'GE65_Tot_Drug_Cst', 'GE65_Tot_Benes', 'LIS_Bene_Cst_Shr', 'NonLIS_Bene_Cst_Shr'], dtype='object')


In [34]:

# check if the col is empty or no 
data['Tot_Prscrbrs'].notna().any()  # → True if column has data

True

In [35]:
# we will check the col catg 
print(data['Tot_Prscrbrs'].unique())
# why we did this to check if the col containe any

[ 691 1167 2617 ... 5130 4542 5240]


In [36]:
# check if the col is empty or no 
data['Tot_Clms'].notna().any()  # → True if column has data

True

In [37]:
# we will check the col catg 
print(data['Tot_Clms'].unique())
# why we did this to check if the col containe any

[  1613   3269  19634 ... 119831  10171   8985]


In [38]:
# check if the col is empty or no 
data['Tot_30day_Fills'].notna().any()  # → True if column has data

True

In [39]:
# we will check the col catg 
print(data['Tot_30day_Fills'].unique())
# why we did this to check if the col containe any

[ 2874.2  5592.  25152.1 ...  1403.   4492.9   867.1]


In [40]:
# check if the col is empty or no 
data['Tot_Drug_Cst'].notna().any()  # → True if column has data

True

In [41]:
# we will check the col catg 
print(data['Tot_Drug_Cst'].unique())
# why we did this to check if the col containe any

[4.43550400e+04 9.79511800e+04 5.29017588e+06 ... 6.46687000e+03
 3.99467000e+03 1.18616000e+03]


In [42]:

# check if the col is empty or no 
data['Tot_Benes'].notna().any()  # → True if column has data

True

In [43]:
# we will check the col catg 
print(data['Tot_Benes'].unique())
# why we did this to check if the col containe any

[  699.  1267.  2807. ...  8316.  9161. 17559.]


In [44]:
# check if the col is empty or no 
data['GE65_Tot_Clms'].notna().any()  # → True if column has data

True

In [45]:
# we will check the col catg 
print(data['GE65_Tot_Clms'].unique())
# why we did this to check if the col containe any

[ 1329.  2567. 11434. ...  9322.  4609.  4930.]


In [46]:

# check if the col is empty or no 
data['GE65_Tot_30day_Fills'].notna().any()  # → True if column has data

True

In [47]:
# we will check the col catg 
print(data['GE65_Tot_30day_Fills'].unique())
# why we did this to check if the col containe any

[ 2375.6  4545.  15353.6 ...  1344.8   393.2  4172.4]


In [48]:

# check if the col is empty or no 
data['GE65_Tot_Drug_Cst'].notna().any()  # → True if column has data

True

In [49]:
# we will check the col catg 
print(data['GE65_Tot_Drug_Cst'].unique())
# why we did this to check if the col containe any

[3.5803140e+04 7.6544140e+04 3.0425690e+06 ... 1.4931649e+05 6.4668700e+03
 7.8558000e+02]


In [50]:

# check if the col is empty or no 
data['GE65_Tot_Benes'].notna().any()  # → True if column has data

True

In [51]:
# we will check the col catg 
print(data['GE65_Tot_Benes'].unique())
# why we did this to check if the col containe any

[  565.  1035.  1662. ...  7402. 16066.  8066.]


In [52]:
# check if the col is empty or no 
data['LIS_Bene_Cst_Shr'].notna().any()  # → True if column has data

True

In [53]:
# we will check the col catg 
print(data['LIS_Bene_Cst_Shr'].unique())
# why we did this to check if the col containe any

[  992.77  1899.59 10643.29 ...   140.42    91.58    12.37]


In [54]:
# we will check the col catg 
print(data['NonLIS_Bene_Cst_Shr'].unique())
# why we did this to check if the col containe any

[  7093.55  10392.03 152140.95 ...   1375.86    408.01    364.9 ]


In [55]:
null = data.isnull().sum()
print(null)


Prscrbr_Geo_Lvl             0
Prscrbr_Geo_Cd           3608
Prscrbr_Geo_Desc            1
Brnd_Name                   0
Gnrc_Name                   0
Tot_Prscrbrs                0
Tot_Clms                    0
Tot_30day_Fills             0
Tot_Drug_Cst                0
Tot_Benes               21596
GE65_Tot_Clms           21063
GE65_Tot_30day_Fills    21063
GE65_Tot_Drug_Cst       21063
GE65_Tot_Benes          48848
LIS_Bene_Cst_Shr            0
NonLIS_Bene_Cst_Shr         0
Opioid_Drug_Flag            0
Opioid_LA_Drug_Flag         0
Antbtc_Drug_Flag            0
Antpsyct_Drug_Flag          0
dtype: int64


- Handling missing values 

| Missing                   | % Missing | Action                                                |
| ------------------------- | --------- | ----------------------------------------------------- |
| `Tot_Benes` = 21,596      | \~14%     | **Fill with 0** or **median**        |
| `GE65_Tot_*` ≈ 21,000     | \~14%     | **Fill with 0** (common for suppressed Medicare data) |
| `GE65_Tot_Benes` = 48,848 | \~33%     | Consider **dropping this column** (too sparse)        |
| `Prscrbr_Geo_Cd` = 3,608  | \~2.4%    | Likely "National" → **Keep as is**                    |
| `Prscrbr_Geo_Desc` = 1    | \~0%      | **Drop the row** or **fill**                          |


In [56]:
# Drop GE65_Tot_Benes (too many missing)
data = data.drop(columns=['GE65_Tot_Benes'])

# Fill remaining GE65_* and Tot_Benes with 0
cols_to_fill = ['Tot_Benes', 'GE65_Tot_Clms', 'GE65_Tot_30day_Fills', 'GE65_Tot_Drug_Cst']
data[cols_to_fill] = data[cols_to_fill].fillna(0)

# Drop the one row with missing Geo_Desc
data = data.dropna(subset=['Prscrbr_Geo_Desc'])


- ✅ Export cleaned dataset to CSV:

In [57]:
data.to_csv('Cleaned_Dataset.csv', index=False)